File: src/main/java/com/vspk/bookrest/security/JwtUserFactory.java
package com.vspk.bookrest.security;

import com.vspk.bookrest.domain.Role;
import com.vspk.bookrest.domain.Status;
import com.vspk.bookrest.domain.User;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class JwtUserFactory {

    private static final List<Status> enabledStatuses = List.of(Status.ACTIVE, Status.VERIFIED);

    public JwtUserFactory() {
    }

    public static JwtUser create(User user) {
        return new JwtUser(
                user.getId(),
                user.getUsername(),
                user.getFirstName(),
                user.getLastName(),
                user.getEmail(),
                user.getPassword(),
                mapToGrantedAuthorities(new ArrayList<>(user.getRoles())),
                enabledStatuses.contains(user.getStatus()),
                user.getUpdated(),
                user.getVerificationTimesAsked()
        );
    }

    private static List<GrantedAuthority> mapToGrantedAuthorities(List<Role> userRoles) {
        return userRoles.stream()
                .map(role ->
                        new SimpleGrantedAuthority(role.getName())
                ).collect(Collectors.toList());
    }
}

File: src/main/java/com/vspk/bookrest/security/JwtUserFactory.kt
package com.vspk.bookrest.security

import com.vspk.bookrest.domain.Role
import com.vspk.bookrest.domain.Status
import com.vspk.bookrest.domain.User
import java.util.stream.Collectors
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority

object JwtUserFactory {

    private val enabledStatuses = listOf(Status.ACTIVE, Status.VERIFIED)

    fun create(user: User): JwtUser {
        return JwtUser(
            user.id!!,
            user.username,
            user.firstName,
            user.lastName,
            user.email,
            user.password,
            mapToGrantedAuthorities(ArrayList(user.roles)),
            enabledStatuses.contains(user.status),
            user.updated!!,
            user.verificationTimesAsked
        )
    }

    private fun mapToGrantedAuthorities(userRoles: List<Role>): List<GrantedAuthority> {
        return userRoles.stream()
            .map { role: Role -> SimpleGrantedAuthority(role.name) }
            .collect(Collectors.toList())
    }
}